packages/opentelemetry-core/src/trace/TraceState.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@opentelemetry/types';
+
+// TODO validate maximum number of items
+// const maximumTraceStateItems = 32;
+
+const maximumTraceStateLength = 512;
+export interface InternalTraceState {
```

Comment:
What would you think about just having `export interface TraceState { [key: string]: string; }`?

In that case there would be no need for the extra class with various wrapper functions, which would add extra weight to an eventual JS bundle for web clients, and adds extra indirection for Node users too. The `serialize` and `deserialize` functions could work in a pretty similar way.

Reply:
Sure, but then how would you preserve the order of key-value pairs? 
As per the [specs](https://w3c.github.io/trace-context/#mutating-the-tracestate-field): 
- New key-value pair should be added into the beginning of the list.
- Modified keys MUST be moved to the beginning of the list.

Reply:
OK, interesting, that seems like a good argument for the extra complexity here. 

Next question: why does `InternalTraceState` need to be exported? What if instead you optionally took a serialized `string` in the constructor that it used to parse?

Reply:
JavaScript objects are not ordered. In general, a specific order doesn't make sense in JavaScript anyway. If this needs to be ordered somehow because of other languages interoperability, then it cannot be a simple object.